### PR TITLE
Vikings issue five

### LINKED ISSUE #5: Images by Tag
#### As a user I want to view all images associated with a tag.

__Story__:

Now that we have added the ability for images to be associated with tags, let
us provide the ability for users to filter images by a given tag. Nana, who
loves fast planes, will be delighted to see only images labeled with SR-71.

__Acceptance criteria__:
- [ ] When I click on a tag, I see a filtered list of only the images that have
  the tag I clicked on.
- [ ] The tag name, not its ID, is used to perform the filtering.

__Discussion topics__:
- [ ] Where in the code should we implement this functionality?
- [ ] What happens if a user tries to manually filter by a nonexistent tag?

__Dependencies__:
- Image Tags

### PR DESCRIPTION
**Pull request by yidanyuan**
Closes #5

---

The following is a checklist to remind you of the steps to take when opening a
new PR. The horizontal line above and everything below can be removed from the
pull request description.

- [ ] I have replaced `X` above with the associated issue number.
- [ ] I have changed the issue's label to "pull request".
- [ ] I have assigned the designated reviewer to this PR.
- [ ] I have set the appropriate label on this PR.
- [ ] I have written clear commit messages ([ref](http://chris.beams.io/posts/git-commit/)).

If needed, please review the document on the
[Image Sharing Project](https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project)
for information on how labels should be used.

### CODE REVIEW

File: test/services/images_tag_filter_test.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+      { url: 'https://66.media.tumblr.com/a09fcb4bda9167908879928ac268f68e/tumblr_pr5mknDjXs1v6ev98o3_1280.jpg',
+        tag_list: %w[tag2 tag3] }
+    ]
+    Image.create!(images)
+    assert_equal ImagesTagFilter.filter(tag: 'tag1').length, 2
+    assert_equal ImagesTagFilter.filter(tag: 'tag2').length, 3
+    assert_equal ImagesTagFilter.filter(tag: 'tag3').length, 2
+    assert_equal ImagesTagFilter.filter(tag: %w[tag3 tag2]).length, 2
+    assert_equal ImagesTagFilter.filter(tag: 'FakeTag').length, 0
+  end
```
Comment (AgentCyrusJr) [inline test/services/images_tag_filter_test.rb]: Could we add a test to check if the params is nil for :tag

File: app/controllers/images_controller.rb [hunk 2]
```diff
@@ -1,6 +1,11 @@
 class ImagesController < ApplicationController
   def index
-    @images = Image.all.order(created_at: :desc)
+    @images = if image_tag_params.nil?
+                Image.all.order(created_at: :desc)
+              else
+                ImagesTagFilter.filter(tag: image_tag_params)
+              end
+    redirect_to images_path if !image_tag_params.nil? && @images.empty?
```
Comment (AgentCyrusJr) [inline app/controllers/images_controller.rb]: We can simplify the logic here a bit. Let us discuss this tomorrow morning

File: app/views/images/index.html.erb [hunk 3]
```diff
@@ -3,12 +3,12 @@
   <h3> <%= link_to 'Upload image using url', new_image_path %> </h3>
   <h2>List of existing images</h2>
   <div>
-    <% @images.each do |f| %> </br>
-      <%= image_tag(f.url, width: '400') %>
-        <% if f.tag_list.any? %>
+    <% @images.each do |img| %> </br>
```
Comment (AgentCyrusJr) [inline app/views/images/index.html.erb]: 👍

File: app/services/images_tag_filter.rb [hunk 4]
```diff
@@ -0,0 +1,9 @@
+class ImagesTagFilter
+  def self.filter(tag:)
+    if tag.nil?
```
Comment (AgentCyrusJr) [inline app/services/images_tag_filter.rb]: Lets remove the nil? here, because it is redundant

Comment (yidanyuan) [reply app/services/images_tag_filter.rb]: Changed to if tag
